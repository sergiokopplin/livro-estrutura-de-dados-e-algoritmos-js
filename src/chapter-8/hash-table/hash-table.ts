import { Dictionary, defaultToString, ValuePair } from '../dictionary/dictionary';

export class HashTable extends Dictionary {
  table: {
    [key: string]: any;
  };
  toStrFn: any;

  constructor(toStrFn = defaultToString) {
    super();
    this.table = {};
    this.toStrFn = toStrFn;
  }

  loseLoseHashCode(key: string) {
    const tableKey: string = this.toStrFn(key);
    let hash = 0;

    for (let i = 0; i < tableKey.length; i++) {
      hash += tableKey.charCodeAt(i);
    }

    return hash % 37;
  }

  djb2HashCode(key: string) {
    const tableKey: string = this.toStrFn(key);

    let hash = 5381;

    for (let i = 0; i < tableKey.length; i++) {
      hash = hash * 33 + tableKey.charCodeAt(i);
    }

    return hash % 1013;
  }

  hashCode(key: string): number {
    return this.djb2HashCode(key);
  }

  put(key: string, value: any): boolean {
    if (key != null && value != null) {
      const tableKey: number = this.djb2HashCode(key);
      this.table[tableKey] = new ValuePair(key, value);

      return true;
    }

    return false;
  }

  get(key: string): string {
    const tableKey: number = this.djb2HashCode(key);
    return this.table[tableKey] ? this.table[tableKey].value : undefined;
  }

  remove(key: string): boolean {
    const tableKey: number = this.djb2HashCode(key);

    if (this.table[tableKey]) {
      delete this.table[tableKey];
      return true;
    }

    return false;
  }
}
