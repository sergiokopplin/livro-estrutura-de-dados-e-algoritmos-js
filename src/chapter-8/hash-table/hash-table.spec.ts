import { HashTable } from './hash-table';

describe('HashTable', () => {
  test('Should construct correctly', () => {
    const hashTable = new HashTable();

    expect(hashTable).toMatchSnapshot();
  });

  test('Should work properly', () => {
    const hashTable = new HashTable();

    hashTable.put('Gandalf', 'gandalf@email.com');
    hashTable.put('John', 'John@email.com');
    hashTable.put('Tyrion', 'Tyrion@email.com');

    expect(hashTable.hashCode('Gandalf')).toBe(798);
    expect(hashTable.hashCode('John')).toBe(838);
    expect(hashTable.hashCode('Tyrion')).toBe(624);
    expect(hashTable.get('Gandalf')).toBe('gandalf@email.com');
    expect(hashTable.get('Kopplin')).toBe(undefined);

    hashTable.remove('Gandalf');
    expect(hashTable.get('Gandalf')).toBe(undefined);
  });
});
