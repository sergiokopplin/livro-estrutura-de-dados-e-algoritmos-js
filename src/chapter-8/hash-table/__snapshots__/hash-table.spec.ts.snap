// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashTable Should construct correctly 1`] = `
HashTable {
  "table": {},
  "toStrFn": [Function],
}
`;
